Assets/Scripts/TextDecoder/NarrativeScript.cs

```
@@ -0,0 +1,107 @@
+    /// <param name="actionDecoder">An optional action decoder, used for testing</param>
+    public NarrativeScript(TextAsset script, DialogueControllerMode type, IActionDecoder actionDecoder = null)
+    {
+        Script = script;
+        Type = type;
+        Initialize(actionDecoder);
+    }
+
+    /// <summary>
+    /// Initialises script values that cannot be set in the Unity inspector
+    /// and begins script reading and object preloading.
+    /// </summary>
+    /// <param name="actionDecoder">An optional action decoder, used for testing</param>
+    public void Initialize(IActionDecoder actionDecoder = null)
+    {
+        _objectStorage = new ObjectStorage();
+        Story = new Story(Script.text);
+        ReadScript(Story, actionDecoder ?? new ObjectPreloader(_objectStorage));
+    }
+
+    /// <summary>
+    /// Creates an action decoder and assigns an ObjectPreloader to its controller properties.
+    /// Gets all lines from an Ink story, extracts all of the action lines (using a
+    /// hash set to remove duplicate lines). Calls the ActionDecoder's OnNewActionLine method
+    /// for each action extracted. ActionDecoder then calls the actions
+    /// on the ObjectPreloader to preload any required assets.
+    /// </summary>
+    /// <param name="story">The Ink story to read</param>
+    /// <param name="actionDecoder">An optional action decoder, used for testing</param>
+    private void ReadScript(Story story, IActionDecoder actionDecoder)
+    {
+        var lines = new List<string>();
+        
+        ReadContent(story.mainContentContainer.content, lines);
+        ReadContent(story.mainContentContainer.namedOnlyContent?.Values.ToList(), lines);
+
+        var actions = new HashSet<string>(lines.Where(line => line[0] == '&'));
+        foreach (var action in actions)
+        {
+            try
+            {
+                actionDecoder.OnNewActionLine(action);
+            }
+            catch (ScriptParsingException exception)
```

Any reason exceptions are silently discarded here? Usually they should only occur, when the script has invalid syntax, at which point I would except all parsers to halt.

---

Having an abstract method for every action means we would need to update three files for every new action added, ActionDecoderBase, ActionDecoder, and ObjectPreloader as abstract methods must be implemented. This kind of goes against the idea of making actions easier to add. Furthermore having all the methods be abstract would mean the ObjectPreloader would still be full of empty methods which is what I thought we wanted to avoid.

Therefore I only included methods that required objects to be loaded, however this leads to ScriptParsingExceptions when loading the objects as the methods are not present on ObjectPreloader, so I put in the try/catch to prevent that, and as its intended functionality I didn't put any warning or error.

Could do some reflection and linq stuff to get available method names, but I feel like this is way more readable.

---

Totally valid point from you with the reusability though. 

Once this is merged, I could add a sub-type, `MethodNotFoundScriptParsingException` which we can handle (and ignore) here, but still have it throw on regular `ScriptParsingException`s. Would that work for you?

---

Yep, that sounds like it could work.

---

Implemented the exception in 22f3be6 and updated this try-catch-block in 86523b0.